lib/Sema/QuoteTransform.cpp

```diff
@@ -0,0 +1,1361 @@
+              DeclRefExpr(ConcreteDeclRef(attr->getQuoteDecl()), DeclNameLoc(),
+                          /*Implicit=*/true);
+          auto quoteBase =
+              TypeExpr::createImplicit(expr->getBase()->getType(), ctx);
+          auto quoteDot =
+              new (ctx) DotSyntaxCallExpr(quoteRef, SourceLoc(), quoteBase);
+          auto quoteCall = CallExpr::createImplicit(ctx, quoteDot, {}, {});
+          auto &tc = TypeChecker::createForContext(ctx);
+          auto type = tc.getTypeOfQuoteExpr(expr->getType(), expr->getLoc());
+          return makeQuote("Unquote", {quotedExpr, quoteCall, quoteType(type)});
+        } else {
+          return quotedExpr;
+        }
+      }
+    } else {
+      return unknownTree(expr);
+    }
+  }
+
+  // TODO(TF-723): Quote dynamic lookup exprs.
+  UNSUPPORTED_EXPR(DynamicLookupExpr)
+
+  // NOTE: There's no point in quoting IDE support artifacts.
+  UNSUPPORTED_EXPR(EditorPlaceholderExpr)
+
+  Expr *visitEnumIsCaseExpr(EnumIsCaseExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote("Is",
+                     {quoteExpr(expr->getSubExpr()),
+                      quoteType(expr->getEnumElement()->getInterfaceType()),
+                      quoteType(expr->getType())});
+  }
+
+  // NOTE: If typechecking encountered errors, there's no point in quoting
+  // anyway because none of that code will make it to runtime.
+  UNSUPPORTED_EXPR(ErrorExpr)
+
+  Expr *visitFloatLiteralExpr(FloatLiteralExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote("FloatLiteral", {quoteFloat(expr->getDigitsText()),
+                                      quoteType(expr->getType())});
+  }
+
+  Expr *visitForcedCheckedCastExpr(ForcedCheckedCastExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote(
+        "ForceAs", {quoteExpr(expr->getSubExpr()), quoteType(expr->getType())});
+  }
+
+  Expr *visitForceTryExpr(ForceTryExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote("ForceTry", {quoteExpr(expr->getSubExpr()),
+                                  quoteType(expr->getType())});
+  }
+
+  Expr *visitForceValueExpr(ForceValueExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote(
+        "Force", {quoteExpr(expr->getSubExpr()), quoteType(expr->getType())});
+  }
+
+  Expr *visitIfExpr(IfExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote("Ternary", {quoteExpr(expr->getCondExpr()),
+                                 quoteExpr(expr->getThenExpr()),
+                                 quoteExpr(expr->getElseExpr()),
+                                 quoteType(expr->getType())});
+  }
+
+  Expr *visitImplicitConversionExpr(ImplicitConversionExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote("Conversion", {quoteExpr(expr->getSubExpr()),
+                                    quoteType(expr->getType())});
+  }
+
+  Expr *visitInOutExpr(InOutExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    if (expr->isImplicit()) {
```

Comment:
What is the general strategy around Exprs with isImplicit? Is this just a one-off case of doing something special in order to make things prettier? Or does some sort of correctness rely on being able to recover fully the original expr before things were rewritten?

Reply:
This is a one-off case because `InOutExpr` seems to be used both syntactically (to represent `&x`) and semantically (to represent certain references to in-out parameters). I've added a comment.